backend/src/main/java/cz/muni/fi/pa165/pneuservis/backend/dao/TireManagerDao.java

```diff
@@ -0,0 +1,79 @@
+package cz.muni.fi.pa165.pneuservis.backend.dao;
+
+import cz.muni.fi.pa165.pneuservis.backend.entity.tire.Manufacturer;
+import cz.muni.fi.pa165.pneuservis.backend.entity.tire.Manufacturer;
+import cz.muni.fi.pa165.pneuservis.backend.entity.tire.Tire;
+import cz.muni.fi.pa165.pneuservis.backend.entity.tire.Tire;
+import cz.muni.fi.pa165.pneuservis.backend.entity.tire.TireProperties;
+import cz.muni.fi.pa165.pneuservis.backend.entity.tire.TireProperties;
+import java.util.List;
+
+/***
+ * @author Jakub Palenik, 422453@mail.muni.cz
+ * 
+ * Interface specifies DAO. When storing the Tire both Manufacturer and TireProperties
+ * are stored if not already in the database. When deleting tire(s) both Manufacturer and 
+ * TireProperties are kept in db. It is expected that these values will be reused 
+ * and are changing rarely.
+ */
+public interface TireManagerDao {
+    
+    /**
+     * Method will store Tire to the database, iff all the variables are properly
+     * initialized. If TireProperties and Manufacturer are not yet in the db
+     * method will store them.
+     * @param tire to be stored 
+     * @return true on success
+     * @throws IllegalArgumentException on tire == null
+     */
+    boolean createTire(Tire tire) throws IllegalArgumentException;
+    
+    /**
+     * Method updates properties of tire old with values of data. Old is to be 
+     * previously stored in db. If old is not found in database method will 
+     * returns false. 
+     * @param old
+     * @param data
+     * @return true on success
+     * @throws IllegalArgumentException on any param == null
+     */
+    boolean updateTire(Tire old, Tire data) throws IllegalArgumentException;
+    
+    /**
+     * Method returns list of all tires in database;
+     * @return 
+     */
+    List<Tire> retrieveAllTires();
+    
+    /**
+     * Method returns list of all tires in database that match the the values 
+     * specified in filter. If filter is null method will return all Tires
+     * @param filter
+     * @return true on success
+     */
+    List<Tire> getTiresByFilter(Tire filter);

```

Comment:
This probably requires a new class that manages the filtering logic. As it is, it wouldn't allow to search for Tires by a field that is null. Let's imagine that I need to want for all Tires that don't have the Manufacturer set, which this wouldn't allow me to.

I suggest creating a class TireFilter that has all the fields that Tire does, but has them as Optional. Then, Optional.empty() can be used to look for fields that are supposed to be null. This example should be much more explanatory:

```java
public class TireManagerDaoImpl {
  public List<Tire> getTiresByFilter(TireFilter filter) {
    SqlStatement sql = sql(); // Just an example of SQL
    // Filtering a field that can't be null
    if (filter.id != null) {
      sql.filter("id", filter.id);
    }
    // Filtering a field that can be null
    if (filter.manufacturer != null) {
      sql.filter("manufacturer", filter.manufacturer.orElse(null));
    }
  }
  
  public class TireFilter {
    private Long id;
    // Manufacturer does not have to be set for a tire
    private Optional<Manufacturer> manufacturer;
  }
  
  public void main(String... args) {
    Filter f = new TireFilter();
    f.manufacturer = Optional.empty();
    // Finds all tires where the manufacturer is not set
    new TireManagerDaoImpl().getTiresByFilter(filter);
  }
}
```